### PR TITLE
[GenericPoseGraph] Add a loss function.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

File: cartographer/pose_graph/constraint/constraint.h [hunk 1]
```diff
@@ -45,8 +48,13 @@ class Constraint {
  protected:
   virtual proto::CostFunction ToCostFunctionProto() const = 0;
 
+  const std::unique_ptr<ceres::LossFunction>& ceres_loss() const {
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/constraint.h]: Returning a std::unique_ptr<>& looks weird. Remove a pointer instead? (To me returning a unique_ptr<> always indicates ownership transfer)

File: cartographer/pose_graph/constraint/loss_function/loss_function_test.cc [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+}
+
+TEST(LossFunctionTest, ConstructHuberLoss) {
+  LossFunction huber_loss(
+      ParseProto<proto::LossFunction>(R"(huber_loss: { scale: 0.5 })"));
+  EXPECT_NE(nullptr,
+            dynamic_cast<ceres::HuberLoss*>(huber_loss.ceres_loss().get()));
+}
+
+TEST(LossFunctionTest, FailToConstructUnspecifiedLoss) {
```
Comment (mbokeloh) [inline cartographer/pose_graph/constraint/loss_function/loss_function_test.cc]: Nit: TEST(LossFunctionDeathTest, ...)

File: cartographer/pose_graph/constraint/constraint.h (hunk 1 above)
Comment (mbokeloh) [reply cartographer/pose_graph/constraint/constraint.h]: I meant 'return a pointer instead'

**Comment by pifon2a** [comment]
@mbokeloh  Done, thanks!

**Comment by pifon2a** [comment]
@wally-the-cartographer merge

**Comment by wally-the-cartographer** [comment]
Merge requested by authorized user pifon2a. Merge queue now has a length of 1.

**Pull request merged**